db/migrate/20190617191109_move_awx_credentials_to_authentications.rb

```diff
@@ -0,0 +1,142 @@
+        update_authentication(auth, awx_info)
+      end
+    end
+  rescue PG::ConnectionBad
+    say("awx database is unreachable, credentials cannot be migrated")
+    return
+  end
+
+  private
+
+  def embedded_ansible_authentications
+    types = %w[
+      ManageIQ::Providers::EmbeddedAnsible::AutomationManager::Credential
+      ManageIQ::Providers::EmbeddedAnsible::AutomationManager::CloudCredential
+      ManageIQ::Providers::EmbeddedAnsible::AutomationManager::MachineCredential
+      ManageIQ::Providers::EmbeddedAnsible::AutomationManager::NetworkCredential
+      ManageIQ::Providers::EmbeddedAnsible::AutomationManager::ScmCredential
+      ManageIQ::Providers::EmbeddedAnsible::AutomationManager::VaultCredential
+      ManageIQ::Providers::EmbeddedAnsible::AutomationManager::AmazonCredential
+      ManageIQ::Providers::EmbeddedAnsible::AutomationManager::AzureCredential
+      ManageIQ::Providers::EmbeddedAnsible::AutomationManager::GoogleCredential
+      ManageIQ::Providers::EmbeddedAnsible::AutomationManager::OpenstackCredential
+      ManageIQ::Providers::EmbeddedAnsible::AutomationManager::RhvCredential
+      ManageIQ::Providers::EmbeddedAnsible::AutomationManager::VmwareCredential
+    ]
+    Authentication.where(:type => types)
+  end
+
+  def awx_credential_info(awx_id)
+    cred_info = awx_connection.async_exec("SELECT inputs FROM main_credential WHERE id = $1::BIGINT", [awx_id]).first
+    # in case there is no matching credential on the awx side
+    return {} unless cred_info
+
+    JSON.parse(cred_info["inputs"])
+  end
+
+  def awx_connection
+    @awx_connection ||= PG::Connection.new(ApplicationRecord.connection.raw_connection.conninfo_hash.merge(:dbname => "awx").delete_blanks)
+  end
+
+  def update_authentication(auth, awx_info)
+    auth.options = auth.options.slice(*OPTIONS_FIELDS.map(&:to_sym))
+    auth.options = nil if auth.options.empty?
+
+    awx_info.each do |k, v|
+      if OPTIONS_FIELDS.include?(k)
+        auth.options ||= {}
+        auth.options[k.to_sym] = v
+        next
+      end
+
+      authentication_attribute = FIELD_MAP[k]
+      decrypted_value          = decrypted_awx_value(k, v, auth.manager_ref)
+      new_value                = ENCRYPTED_ATTRIBUTES.include?(authentication_attribute) ? ManageIQ::Password.encrypt(decrypted_value) : decrypted_value
+
+      if authentication_attribute
+        auth.send("#{authentication_attribute}=", new_value)
+      else
+        say("Unknown credential field #{k}, ignoring")
+      end
+    end
+    auth.save!
+  end
+
+  def decrypted_awx_value(field, value, pk)
+    require 'awesome_spawn'
+
+    return value unless value.include?("$encrypted$")
+    env = {
+      "FIELD_NAME"  => field,
+      "VALUE"       => value,
+      "SECRET_KEY"  => secret_key,
+      "PRIMARY_KEY" => pk
+    }
+    result = AwesomeSpawn.run("python3", :params => [script_path], :env => env)
+
+    raise "Failed to decrypt #{field} for credential #{pk}" if result.failure?
```

Comment:
Should we just use `run!` since we don't actually expect this expectation?

Reply:
Given that we have passwords in plaintext in the ENV hash I wanted tighter control of the error message.